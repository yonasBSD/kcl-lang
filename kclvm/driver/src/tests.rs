use std::path::PathBuf;
use std::{env, fs, panic};

use kclvm_config::modfile::get_vendor_home;
use kclvm_config::settings::KeyValuePair;
use kclvm_parser::LoadProgramOptions;
use walkdir::WalkDir;

use crate::arguments::parse_key_value_pair;
use crate::toolchain::Toolchain;
use crate::toolchain::{fill_pkg_maps_for_k_file, CommandToolchain, NativeToolchain};
use crate::{get_pkg_list, lookup_the_nearest_file_dir, toolchain};

#[test]
fn test_parse_key_value_pair() {
    let cases = [
        (
            "k=v",
            KeyValuePair {
                key: "k".to_string(),
                value: "\"v\"".into(),
            },
        ),
        (
            "k=1",
            KeyValuePair {
                key: "k".to_string(),
                value: "1".into(),
            },
        ),
        (
            "k=None",
            KeyValuePair {
                key: "k".to_string(),
                value: "null".into(),
            },
        ),
        (
            "k=True",
            KeyValuePair {
                key: "k".to_string(),
                value: "true".into(),
            },
        ),
        (
            "k=true",
            KeyValuePair {
                key: "k".to_string(),
                value: "true".into(),
            },
        ),
        (
            "k={\"key\": \"value\"}",
            KeyValuePair {
                key: "k".to_string(),
                value: "{\"key\": \"value\"}".into(),
            },
        ),
        (
            "k=[1, 2, 3]",
            KeyValuePair {
                key: "k".to_string(),
                value: "[1, 2, 3]".into(),
            },
        ),
    ];
    for (value, pair) in cases {
        let result = parse_key_value_pair(value).unwrap();
        assert_eq!(result.key, pair.key);
        assert_eq!(result.value, pair.value);
    }
}

fn clear_path(path: PathBuf) {
    WalkDir::new(path)
        .into_iter()
        .filter_map(|e| e.ok())
        .for_each(|e| {
            fs::remove_file(e.path())
                .or_else(|_| fs::remove_dir(e.path()))
                .ok();
        });
}

#[test]
fn test_parse_key_value_pair_fail() {
    let cases = ["=v", "k=", "="];
    for case in cases {
        assert!(parse_key_value_pair(case).is_err());
    }
}

fn test_fill_pkg_maps_for_k_file_with_line() {
    let root_path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_metadata_with_line");

    let main_pkg_path = root_path.join("main_pkg").join("main.k");
    let dep_with_line_path = root_path.join("dep-with-line");

    let mut opts = LoadProgramOptions::default();
    assert_eq!(format!("{:?}", opts.package_maps), "{}");

    let res = fill_pkg_maps_for_k_file(&toolchain::default(), main_pkg_path.clone(), &mut opts);
    assert!(res.is_ok());

    let pkg_maps = opts.package_maps.clone();
    assert_eq!(pkg_maps.len(), 1);
    assert!(pkg_maps.get("dep_with_line").is_some());

    assert_eq!(
        PathBuf::from(pkg_maps.get("dep_with_line").unwrap().clone())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        dep_with_line_path
            .canonicalize()
            .unwrap()
            .display()
            .to_string()
    );
}

#[test]
fn test_native_fill_pkg_maps_for_k_file_with_line() {
    let root_path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_metadata_with_line");

    let main_pkg_path = root_path.join("main_pkg").join("main.k");
    let dep_with_line_path = root_path.join("dep-with-line");

    let mut opts = LoadProgramOptions::default();
    assert_eq!(format!("{:?}", opts.package_maps), "{}");

    let res = fill_pkg_maps_for_k_file(
        &NativeToolchain::default(),
        main_pkg_path.clone(),
        &mut opts,
    );
    assert!(res.is_ok());

    let pkg_maps = opts.package_maps.clone();
    assert_eq!(pkg_maps.len(), 1);
    assert!(pkg_maps.get("dep_with_line").is_some());

    assert_eq!(
        PathBuf::from(pkg_maps.get("dep_with_line").unwrap().clone())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        dep_with_line_path
            .canonicalize()
            .unwrap()
            .display()
            .to_string()
    );
}

fn test_fill_pkg_maps_for_k_file() {
    let path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_metadata")
        .join("subdir")
        .join("main.k");

    let vendor_path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("test_vendor");

    env::set_var(
        "KCL_PKG_PATH",
        vendor_path.canonicalize().unwrap().display().to_string(),
    );

    let mut opts = LoadProgramOptions::default();
    assert_eq!(format!("{:?}", opts.package_maps), "{}");

    let res = fill_pkg_maps_for_k_file(&toolchain::default(), path.clone(), &mut opts);
    assert!(res.is_ok());
    let vendor_home = get_vendor_home();

    let pkg_maps = opts.package_maps.clone();
    assert_eq!(pkg_maps.len(), 1);
    assert!(pkg_maps.get("kcl4").is_some());
    assert_eq!(
        PathBuf::from(pkg_maps.get("kcl4").unwrap().clone())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        PathBuf::from(vendor_home)
            .join("kcl4_v0.0.1")
            .canonicalize()
            .unwrap()
            .display()
            .to_string()
    );

    clear_path(vendor_path.join(".kpm"))
}

#[test]
fn test_lookup_the_nearest_file_dir() {
    let path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_metadata");
    let result = lookup_the_nearest_file_dir(path.clone(), "kcl.mod");
    assert!(result.is_some());
    assert_eq!(
        result.unwrap().display().to_string(),
        path.canonicalize().unwrap().display().to_string()
    );

    let main_path = path.join("subdir").join("main.k");
    let result = lookup_the_nearest_file_dir(main_path, "kcl.mod");
    assert!(result.is_some());
    assert_eq!(
        result.unwrap().display().to_string(),
        path.canonicalize().unwrap().display().to_string()
    );
}

#[test]
fn test_fetch_metadata_in_order() {
    test_cmd_tool_fetch_metadata();
    println!("test_cmd_tool_fetch_metadata() passed");
    test_native_tool_fetch_metadata();
    println!("test_native_tool_fetch_metadata() passed");
    test_fill_pkg_maps_for_k_file();
    println!("test_fill_pkg_maps_for_k_file() passed");
    test_native_fill_pkg_maps_for_k_file_with_line();
    println!("test_native_fill_pkg_maps_for_k_file_with_line() passed");
    test_fill_pkg_maps_for_k_file_with_line();
    println!("test_fill_pkg_maps_for_k_file_with_line() passed");
    test_native_update_dependencies();
    println!("test_native_update_dependencies() passed");
    test_update_dependencies();
    println!("test_update_dependencies() passed");
}

fn test_cmd_tool_fetch_metadata() {
    test_tool_fetch_metadata(CommandToolchain::default())
}

fn test_native_tool_fetch_metadata() {
    test_tool_fetch_metadata(NativeToolchain::default())
}

fn test_tool_fetch_metadata(tool: impl Toolchain) {
    let path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_metadata");

    let vendor_path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("test_vendor");

    env::set_var(
        "KCL_PKG_PATH",
        vendor_path.canonicalize().unwrap().display().to_string(),
    );
    let vendor_home = get_vendor_home();
    let metadata = tool.fetch_metadata(path.clone());
    let pkgs = metadata.unwrap().packages.clone();
    assert_eq!(pkgs.len(), 1);
    assert!(pkgs.get("kcl4").is_some());
    assert_eq!(pkgs.get("kcl4").unwrap().name, "kcl4");
    assert_eq!(
        pkgs.get("kcl4")
            .unwrap()
            .manifest_path
            .clone()
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        PathBuf::from(vendor_home)
            .join("kcl4_v0.0.1")
            .canonicalize()
            .unwrap()
            .display()
            .to_string()
    );
    clear_path(vendor_path.join(".kpm"))
}

#[test]
fn test_fetch_metadata_invalid() {
    let result = panic::catch_unwind(|| {
        let tool = toolchain::default();
        let result = tool.fetch_metadata("invalid_path".to_string().into());
        match result {
            Ok(_) => {
                panic!("The method should not return Ok")
            }
            Err(_) => {
                println!("return with an error.")
            }
        }
    });

    match result {
        Ok(_) => println!("no panic"),
        Err(e) => panic!("The method should not panic forever.: {:?}", e),
    }
}

#[test]
fn test_native_fetch_metadata_invalid() {
    let result = panic::catch_unwind(|| {
        let tool = NativeToolchain::default();
        let result = tool.fetch_metadata("invalid_path".to_string().into());
        match result {
            Ok(_) => {
                panic!("The method should not return Ok")
            }
            Err(_) => {
                println!("return with an error.")
            }
        }
    });

    match result {
        Ok(_) => println!("no panic"),
        Err(e) => panic!("The method should not panic forever.: {:?}", e),
    }
}

#[test]
fn test_get_pkg_list() {
    assert_eq!(get_pkg_list("./src/test_data/pkg_list/").unwrap().len(), 1);
    assert_eq!(
        get_pkg_list("./src/test_data/pkg_list/...").unwrap().len(),
        3
    );
}

fn test_update_dependencies() {
    let path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_update");

    let tool = toolchain::default();
    let update_mod = tool.update_dependencies(path.clone());
    // Show more information when the test fails.
    println!("{:?}", update_mod);
    assert!(update_mod.is_ok());
}

fn test_native_update_dependencies() {
    let path = PathBuf::from(".")
        .join("src")
        .join("test_data")
        .join("kpm_update");

    let tool = NativeToolchain::default();
    tool.update_dependencies(path.clone()).unwrap();
}
