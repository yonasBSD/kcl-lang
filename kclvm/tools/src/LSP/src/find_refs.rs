use crate::to_lsp::lsp_location;
use kclvm_error::Position as KCLPos;
use kclvm_sema::core::global_state::GlobalState;
use lsp_types::Location;
use std::collections::HashSet;

pub fn find_refs(kcl_pos: &KCLPos, gs: &GlobalState) -> Option<Vec<Location>> {
    match gs.look_up_exact_symbol(kcl_pos) {
        Some(symbol_ref) => match gs.get_symbols().get_symbol(symbol_ref) {
            Some(symbol) => match symbol.get_definition() {
                Some(def_ref) => {
                    if let Some(def) = gs.get_symbols().get_symbol(def_ref) {
                        let refs = def.get_references();
                        let mut refs_locs: HashSet<(KCLPos, KCLPos)> = refs
                            .iter()
                            .filter_map(|symbol| {
                                gs.get_symbols()
                                    .get_symbol(*symbol)
                                    .map(|sym| sym.get_range())
                            })
                            .collect();
                        refs_locs.insert(symbol.get_range());
                        refs_locs.insert(def.get_range());
                        let mut res: Vec<Location> = refs_locs
                            .iter()
                            .filter_map(|(start, end)| {
                                lsp_location(start.filename.clone(), &start, &end).map(|loc| loc)
                            })
                            .collect();
                        res.sort_by_key(|e| e.range.start.line);
                        return Some(res);
                    }
                }
                None => {}
            },
            None => {}
        },
        None => {}
    };
    None
}

#[cfg(test)]
mod tests {
    use crate::find_refs::find_refs;
    use crate::from_lsp::file_path_from_url;
    use lsp_types::Location;
    use std::path::{Path, PathBuf};

    use crate::tests::compile_test_file;
    use kclvm_error::Position as KCLPos;

    #[macro_export]
    macro_rules! find_ref_test_snapshot {
        ($name:ident, $file:expr, $line:expr, $column: expr) => {
            #[test]
            fn $name() {
                let (file, _program, _, gs) = compile_test_file($file);

                let pos = KCLPos {
                    filename: file.clone(),
                    line: $line,
                    column: Some($column),
                };
                let res = find_refs(&pos, &gs);
                insta::assert_snapshot!(format!("{}", { fmt_resp(&res) }));
            }
        };
    }

    fn fmt_resp(resp: &Option<Vec<Location>>) -> String {
        let root_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        match resp {
            Some(resp) => {
                let mut res = String::new();
                for loc in resp {
                    let url = file_path_from_url(&loc.uri).unwrap();
                    let got_path = Path::new(&url);
                    let relative_path = got_path.strip_prefix(root_path.clone()).unwrap();
                    res.push_str(&format!(
                        "path: {:?}, range: {:?}\n",
                        relative_path, loc.range
                    ));
                }
                res
            }
            None => "None".to_string(),
        }
    }

    find_ref_test_snapshot!(
        find_refs_variable_def_test,
        "src/test_data/find_refs_test/main.k",
        1,
        1
    );

    find_ref_test_snapshot!(
        find_refs_variable_ref_test,
        "src/test_data/find_refs_test/main.k",
        2,
        5
    );

    find_ref_test_snapshot!(
        find_refs_schema_name_test,
        "src/test_data/find_refs_test/main.k",
        5,
        8
    );

    find_ref_test_snapshot!(
        find_refs_schema_name_ref_test,
        "src/test_data/find_refs_test/main.k",
        9,
        8
    );

    find_ref_test_snapshot!(
        find_refs_schema_attr_test,
        "src/test_data/find_refs_test/main.k",
        6,
        7
    );

    find_ref_test_snapshot!(
        find_refs_schema_attr_ref_test,
        "src/test_data/find_refs_test/main.k",
        13,
        11
    );

    find_ref_test_snapshot!(
        find_refs_schema_arg_test,
        "src/test_data/find_refs_test/main.k",
        17,
        17
    );

    find_ref_test_snapshot!(
        find_refs_schema_arg_1_test,
        "src/test_data/find_refs_test/main.k",
        18,
        17
    );
}
