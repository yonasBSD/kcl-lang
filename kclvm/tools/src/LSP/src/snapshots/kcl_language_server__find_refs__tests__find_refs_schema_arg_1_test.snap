---
source: tools/src/LSP/src/find_refs.rs
expression: "format!(\"{}\", { fmt_resp(& res) })"
---
path: "src/test_data/find_refs_test/main.k", range: Range { start: Position { line: 16, character: 14 }, end: Position { line: 16, character: 19 } }
path: "src/test_data/find_refs_test/main.k", range: Range { start: Position { line: 17, character: 13 }, end: Position { line: 17, character: 18 } }
